File: Java/Graphs/ProblemsOnDfsBfs/1073-number-of-enclaves/number-of-enclaves.java
class Solution {
    private void dfs(int row,int col,int grid[][],boolean visited[][], int delRow[],int delCol[]){
        visited[row][col] = true;

        for(int i=0;i<4;i++){
            int nrow = row + delRow[i];
            int ncol = col + delCol[i];

            if(nrow >=0 && nrow < grid.length && ncol >= 0 && ncol < grid[0].length && !visited[nrow][ncol] && grid[nrow][ncol] == 1 ){
                dfs(nrow,ncol,grid,visited,delRow,delCol);
            }
        }
    }
    public int numEnclaves(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        boolean visited[][] = new boolean[m][n];

        int delRow[] = {-1,0,1,0};
        int delCol[] = {0,1,0,-1};
        // col
        for(int j = 0;j<n;j++){
            
            if(!visited[0][j] && grid[0][j] == 1){
                dfs(0,j,grid,visited,delRow,delCol);
            }

            if(!visited[m-1][j] && grid[m-1][j] == 1){
                dfs(m-1,j,grid,visited,delRow,delCol);
            }
        }

        // row
        for(int i=0;i<m;i++){
            
            if(!visited[i][0] && grid[i][0] == 1){
                dfs(i,0,grid,visited,delRow,delCol);
            }

            if(!visited[i][n-1] && grid[i][n-1] == 1){
                dfs(i,n-1,grid,visited,delRow,delCol);
            }
        }

        int cnt = 0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(!visited[i][j] && grid[i][j] == 1){
                    cnt++;
                }
            }
        }
        return cnt;
    }
}

File: Python/Graphs/ProblemsOnBFSAndDFS/1073-number-of-enclaves/number-of-enclaves.py
class Pair:
        def __init__(self, first: int, second: int):
            self.first = first
            self.second = second
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
            
        visited = [[False] * n for _ in range(m)]
            
        delRow = [-1, 0, 1, 0]
        delCol = [0, 1, 0, -1]
            
        q = deque()
            # corner row, col
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and (i == 0 or j == 0 or i == m - 1 or j == n - 1):
                    q.append(Pair(i, j))
                    visited[i][j] = True
            
        while q:
            row = q[0].first
            col = q[0].second
            q.popleft()
                
            for i in range(4):
                nrow = row + delRow[i]
                ncol = col + delCol[i]
                if nrow >= 0 and nrow < m and ncol >= 0 and ncol < n and not visited[nrow][ncol] and grid[nrow][ncol] == 1:
                    visited[nrow][ncol] = True
                    q.append(Pair(nrow, ncol))
            
        cnt = 0
        for i in range(m):
            for j in range(n):
                if not visited[i][j] and grid[i][j] == 1:
                    cnt += 1
        return cnt